xcdat/temporal_avg.py

```diff
@@ -0,0 +1,988 @@
+
+        # Time complexity of O(n^2), but okay with these small data structures.
+        seasons_map = {}
+        for mon_int, mon_str in MONTH_INT_TO_STR.items():
+            for season in custom_seasons:
+                if mon_str in season:
+                    seasons_map[mon_int] = season
+
+        df = df_season.copy()
+        df["season"] = df["month"].map(seasons_map)
+        return df
+
+    def _shift_decembers(self, df_season: pd.DataFrame) -> pd.DataFrame:
+        """Shifts Decembers over to the next year for "DJF" seasons.
+
+        For "DJF" seasons, Decembers must be shifted over to the next year in
+        order for the xarray groupby operation to perform the correct grouping.
+        Otherwise, grouping is performed with the native xarray "DJF" season
+        (which is actually "JFD").
+
+        Parameters
+        ----------
+        df_season : pd.DataFrame
+            The DataFrame of time coordinates using the "season" frequency.
+
+        Returns
+        -------
+        pd.DataFrame
+            The DataFrame of time coordinates with Decembers shifted over to the
+            next year.
+
+        Examples
+        --------
+
+        Comparison of "DJF" and "JFD" seasons:
+
+        >>> # "DJF" (shifted Decembers)
+        >>> [(2000, "DJF", 1), (2000, "DJF", 2), (2001, "DJF", 12),
+        >>>  (2001, "DJF", 1), (2001, "DJF", 2), (2002, "DJF", 12)]
+
+        >>> # "JFD" (native xarray behavior)
+        >>> [(2000, "DJF", 1), (2000, "DJF", 2), (2000, "DJF", 12),
+        >>>  (2001, "DJF", 1), (2001, "DJF", 2), (2001, "DJF", 12)]
+        """
+        df_season.loc[df_season["month"] == 12, "year"] = df_season["year"] + 1
+        return df_season
+
+    def _drop_obsolete_columns(self, df_season: pd.DataFrame) -> pd.DataFrame:
+        """Drops obsolete columns from the time coordinates DataFrame.
+
+        Additional processing is performed on the time coordinates DataFrame
+        for the "season" frequency (refer to ``_process_season_dataframe()``).
+        Processing requires additonal columns related to the time coordinates,
+        which become obsolete after it is done. The obsolete columns must be
+        dropped from the DataFrame before creating the time MultiIndex.
+        Otherwise, the time MultiIndex will include additional levels,
+        resulting in incorrect grouping outputs.
+
+        Parameters
+        ----------
+        df_season : pd.DataFrame
+            The DataFrame of time coordinates for the "season" frequency with
+            obsolete columns.
+
+        Returns
+        -------
+        pd.DataFrame
+            The DataFrame of time coordinates for the "season" frequency with
+            obsolete columns dropped.
+        """
+        if self._mode == "timeseries":
+            df_season = df_season.drop("month", axis=1)
+        elif self._mode in ["climatology", "departure"]:
+            df_season = df_season.drop(["year", "month"], axis=1)
```

Comment:
This is a note to myself to come back to this...it is not clear to me right now why you would drop the year if the mode is departure.

Reply:
Got it. This is only called from the `_process_season_dataframe` function.